### ISSUE TITLE
Issue Accessing REDCap on Custom Domain via Route53 After Successful Deployment

### ISSUE THREAD

**Issue by tomotsuguseki** [aws-samples/redcap-on-aws#16]
Hi,

I'm excited to share that I've successfully deployed my REDCap production environment on AWS using `yarn deploy --stage prod`. 
A big thanks to AWS staffs for their support on this project.

However, I've encountered a new challenge. While I can access our REDCap instance through the AWS provided URL `https://hogehoge.us-east-1.awsapprunner.com`, I'm unable to reach it via our custom domain `redcap.my-domain.com` configured in Route53.

To give you a better understanding, here's an excerpt from our `stages.ts` configuration:

```
import { Cpu, Memory } from '@aws-cdk/aws-apprunner-alpha';
import { Duration } from 'aws-cdk-lib';
import { DomainAppsConfig, RedCapConfig } from './prototyping';

const baseOptions = {
  name: 'REDCap',
  profile: 'default',
  region: 'us-east-1',
  allowedIps: ['my-ip-address'],
};

const prod: RedCapConfig = {
  ...baseOptions,
  phpTimezone: 'Asia/Tokyo',
  redCapS3Path: 'my-s3-repository/redcap14.1.6.zip',
  domain: 'my-domain.com',
  subDomain: 'redcap.my-domain.com',
  hostInRoute53: true,
  email: 'my-email-address',
  appRunnerConcurrency: 10,
  appRunnerMaxSize: 5,
  appRunnerMinSize: 1,
  cronSecret: 'prodsecret',
  cpu: Cpu.TWO_VCPU,
  memory: Memory.FOUR_GB,
  ec2ServerStack: {
    ec2StackDuration: Duration.hours(3),
  },
};

// Default route53NS config, no records are created.
const route53NS: DomainAppsConfig = {
  ...baseOptions,
  profile: 'default',
  region: 'us-east-1',
  apps: [],
  domain: 'my-domain.com',
  subDomain: 'redcap.my-domain.com',
};

export { dev, prod, route53NS, stag };
```

Does anyone have any suggestions or insights on how I might resolve the domain access issue?
Thank you in advance for your time and help!

**Comment by jl4nz** [comment]
Hi, congrats on deploying this project.

I understand the DNS configuration is a bit tricky. The first thing to say, is that `redcap.my-domain.com` is just a sample domain name, it's not real and it won't work. You have to **own** a domain name, you can register your own in Amazon Route53 or other DNS provider.

This project has automatic features like Amazon SES email validation and Certificate generation for SSL if the domain is hosted in Amazon Route 53. It's also possible to use other DNS providers, but it requires more manual intervention to create certificates and validate the setup for REDCap with that domain. 

Last, this configuration in the `stages.ts` , `const route53NS: DomainAppsConfig = { .... } )` ,  is only to use a registered domain name in an `external` AWS account. For example: your company has already registered a domain with AWS, but is not in the same account where you are deploying REDCap. 

We are planing a v1.0.1 release where updated documentation about this + ECS on Fargate support, we hope to do it soon.

So, my question the questions help you better are: Do you have/own a domain for your deployment? and where is it hosted?

**Comment by tomotsuguseki** [comment]
Thank you for your prompt reply.
I have a domain `tomotsuguseki.com` in Route53 in the same account where I deployed REDCap.
Again, how can I do to solve the problem?

**Comment by jl4nz** [comment]
Understand. This case needs an update that we are going to release in the next release. 

However, I can propose the following to try making it work now:

1. Replace these two files: `/stacks/Backend.ts` and `prototyping.d.ts` included in the attached sst.zip file.

[sst.zip](https://github.com/aws-samples/redcap-on-aws/files/14444907/sst.zip)

2. Your stage file should be updated to be something like this:

```ts
const prod: RedCapConfig = {
  ...baseOptions,
  domain: 'tomotsuguseki.com',
  subdomain: 'redcap',
  hostInRoute53: 'tomotsuguseki.com', // <-- important change
  ....
};
```

This will make sure to use your existing domain and hosted zone of Amazon Route 53.

3. Deploy again. `yarn deploy --stage ... `

4. After the deployment and some extra time, the certificate and your domain should work.

Let me know if this works.
Thanks!

**Comment by jl4nz** [comment]
The update should have re-created the appRunner instances. But according to this:

```
   prod-REDCap-service/apprunner-sub: Cannot delete a subscription which is pending confirmation. Detaching subscription from stack.
```
The deploy has been reverted.

Check you email inbox and confirm this subscription and try deploy again.

If this fails, check in the AWS Console / Cloudformation / Backend stack / Events. A list of events is displayed here, in red color a more detailed error can help us to diagnose what is happening.

**Comment by jl4nz** [comment]
I'll try to re-create your environment and see if I can replicate the error... give me a few minutes please.

**Comment by jl4nz** [comment]
Does your AWS console shows repeated  hosted zones names in here?
https://console.aws.amazon.com/route53/v2/hostedzones

**Comment by tomotsuguseki** [comment]
Yes, there are two records in the hosted zone, although I don't understand why.
Did it cause problem?

**Comment by tomotsuguseki** [comment]
Thanks a lot for your help.
I could deploy the app through my custom domain, but I have one more question.
I intended to deploy my REDCap at the subdomain `redcap.tomotsuguseki.com` rather than `tomotsuguseki.com`.
What is necessary for the purpose?

**Comment by jl4nz** [comment]
Glad it "almost" worked. Let's see:

1. Can you confirm the stage config looks like this:

```ts
const prod: RedCapConfig = {
  ...baseOptions,
  domain: 'tomotsuguseki.com',
  subdomain: 'redcap',
  hostInRoute53: 'tomotsuguseki.com',
  ....
};
```

2. At the https://us-east-1.console.aws.amazon.com/route53/v2/hostedzones, click on your hosted zone, there should be a `Value/Route traffic to` column, here one of these records look like `.xxxxxx.awsapprunner.com`, what is the `Type` and `Record name` columns for this row?

**Comment by tomotsuguseki** [comment]
I checked my environment as follows:
1. Stage config seems OK.
2. Type: A; Record name: tomotsuguseki.com

Thanks for your help.

**Comment by jl4nz** [comment]
OK, then there's the problem. I'm not sure why it did an A record, with that setting it should be a CNAME to your subdomain.

Let's try to re-configure this.

1. In the Backend.ts file, can you comment or delete **only** this section of the code

```ts
  // Automatic AppRunner custom domain enablement only if using Route53
  if (domain && publicHostedZone) {
    const DomainName = subdomain ? `${subdomain}.${domain}` : domain;
    const RecordSetType = subdomain ? 'CNAME' : 'A';
    new cloudformation_include.CfnInclude(
      stack,
      `${app.stage}-${app.name}-apprunner-custom-domain`,
      {
        templateFile: './prototyping/cfn/AppRunnerCustomDomain.yaml',
        parameters: {
          DomainName,
          RecordSetType,
          AppRunnerHostedZones: getAppRunnerHostedZone(stack.region),
          ServiceUrl: redCapRunner.service.serviceUrl,
          ServiceArn: redCapRunner.service.serviceArn,
          DNSDomainId: publicHostedZone.hostedZoneId,
        },
      },
    );
    stack.addOutputs({
      RedcapCustomUrl: `https://${DomainName}`,
    });
  }
  ```
  
2. Deploy
  
3. In the AWS console, make sure that AWS App Runner does not have any custom domain
<img width="775" alt="Screenshot 2024-03-01 at 11 43 26" src="https://github.com/aws-samples/redcap-on-aws/assets/94814971/6b7ac79d-5b3c-41c2-a376-f2b2e07aba2e">

4. Revert the code changes in `Backend.ts` to re-configure the custom domain

5. Deploy again

**Comment by tomotsuguseki** [comment]
Thanks for your instruction.
I tried above procedure, but the result was same.
Are any changes necessary?

**Comment by jl4nz** [comment]
I see. The only way I see this is not working can be the subdomain is not set. 

At the start of this issue there is this setting: `  subDomain: 'redcap.my-domain.com', ` , that is wrong. It should be `subdomain: redcap` (lowercap, subdomain only)

**Comment by jl4nz** [comment]
We notice the `subDomain` as a documentation error, will be fix in the next release

**Comment by jl4nz** [comment]
Hi, just for your information, we just released the new version v1.0.1. Please make sure to follow the Upgrade instructions in the `CHANGELOG` file.

Have a nice weekend!
